### PR TITLE
Feature/track drag drop

### PR DESCRIPTION
**Pull request by wolfbiter**
add react-dnd. add PrimaryTrackTableRow, which is both a drag and drop target. add reorderPrimaryTrack and swapPrimaryTracks to mixes reducer

primary track drag drop

### CODE REVIEW

File: layout/container.js [hunk 1]
```diff
@@ -18,4 +20,4 @@ class LayoutContainer extends React.Component {
   }
 }
 
-module.exports = LayoutContainer
+module.exports = DragDropContext(HTML5Backend)(LayoutContainer)
```
Comment (wolfbiter) [inline layout/container.js:23]: is this the right place to add? needs to 'wrap your app component'. this is the highest level place i could find for that, other than routes.js

File: mixes/components/primary-track-row.js [hunk 2]
```diff
@@ -0,0 +1,107 @@
+const React = require('react')
```
Comment (wolfbiter) [inline mixes/components/primary-track-row.js:1]: feel free to ignore this file... it's almost entirely about getting `react-table` and `react-dnd` to play nice together.

the important part is it exposes an API `createPrimaryTrackRow({ onDrop, onDragEnter })` which is used by `PrimaryTrackTable` to generate primary track rows

File: mixes/components/primary-track-table.js [hunk 3]
```diff
@@ -24,7 +38,7 @@ class PrimaryTrackTable extends React.Component {
       render: props => <span title={props.value}>{props.value}</span>
     }, {
       header: 'Key',
-      accessor: 'meta.key',
+      accessor: 'channel.startBeat',
```
Comment (wolfbiter) [inline mixes/components/primary-track-table.js]: TODO: undo this

File: mixes/components/primary-track-table.js [hunk 4]
```diff
@@ -1,16 +1,30 @@
...
+        reorderPrimaryTrack({ targetIndex, id: sourceRowId, sourceIndex, tracks })
+      },
+    })
+
     const columns = [{
       id: 'number',
       header: '#',
-      accessor: (track) => track.index + 1,
+      accessor: track => track.index,
+      render: props => <span title={props.value}>{props.value + 1}</span>,
```
Comment (wolfbiter) [inline mixes/components/primary-track-table.js]: TODO: undo this change

File: mixes/containers/mix.js [hunk 5]
```diff
@@ -50,10 +50,12 @@ class MixContainer extends React.Component {
         placeholder='Untitled Mix'
         onChange={this.onChangeMixTitle.bind(this)} />
 
+    // TODO: replace with react-dnd equivalent
+      //<FileDrop
+      //  frame={document}
+      //  onFrameDrop={this.onFilesDrop.bind(this)} />
```
Comment (wolfbiter) [inline mixes/containers/mix.js]: TODO

File: mixes/reducer.js [hunk 6]
```diff
@@ -171,6 +175,56 @@ function createReducer (config) {
         Effects.constant(navigateToMix(newMix.id))
       ]))
     },
+    [reorderPrimaryTrack]: (state, action) => {
```
Comment (wolfbiter) [inline mixes/reducer.js:177]: this function is complex because we need to be REALLY careful about how we move around channels in our mix. specifically, when moving tracks, we want to move it WITH its trailing transition.

note this is distinct from a normal list because we cant just have an `order` property - we calculate that from `startBeat`

File: mixes/reducer.js [hunk 7]
```diff
@@ -171,6 +175,56 @@ function createReducer (config) {
...
+
+      let effects = []
+
+      // forwards swap
+      if (sourceIndex < targetIndex) {
+        for (let i = sourceIndex; i < targetIndex; i++) {
+
+          // TODO: this does not work.
+          // we need to get the result of previous swap in order to do the next swap
+          // could do that by just passing IDs, and moving swapPrimaryTracks to channel reducer?
```
Comment (wolfbiter) [inline mixes/reducer.js]: TODO

File: mixes/reducer.js [hunk 8]
```diff
@@ -171,6 +175,56 @@ function createReducer (config) {
...
+            sourceTrack: tracks[i],
+            targetTrack: tracks[i + 1]
+          })))
+        }
+      }
+
+      // backwards swap
+      else {
+        // TODO: implement this
+        console.log("BACKWARDS SWAP not implemented")
```
Comment (wolfbiter) [inline mixes/reducer.js]: TODO

File: mixes/reducer.js [hunk 9]
```diff
@@ -171,6 +175,56 @@ function createReducer (config) {
...
+      }
+
+      return loop(state, Effects.batch(effects))
+    },
+    [swapPrimaryTracks]: (state, action) => {
+      // TODO: update both channels, switch startBeats
+      // TODO: also update both associate transitions to same endBeat as track
+      // TODO: anything we need to do with track or transition length?
```
Comment (wolfbiter) [inline mixes/reducer.js]: TODO

File: styles/file-drop.css [hunk 10]
```diff
@@ -16,3 +16,7 @@
     background-color: rgba(0,155,0,0.5);
   }
 }
+
+.bg-green {
+  background-color: green !important;
+}
```
Comment (wolfbiter) [inline styles/file-drop.css]: TODO: update styling. dont leave this.

File: styles/file-drop.css [hunk 11]
```diff
@@ -16,3 +16,11 @@
     background-color: rgba(0,155,0,0.5);
   }
 }
+
+.primary-track-drag-over {
```
Comment (wolfbiter) [inline styles/file-drop.css]: @ahdinosaur starting to need some styling. we should setup that infrastructure next time around

File: mixes/containers/mix.js [hunk 12]
```diff
@@ -50,10 +50,12 @@ class MixContainer extends React.Component {
         placeholder='Untitled Mix'
         onChange={this.onChangeMixTitle.bind(this)} />
 
+    // TODO: replace with react-dnd equivalent
+      // <FileDrop
+      //  frame={document}
+      //  onFrameDrop={this.onFilesDrop.bind(this)} />
```
Comment (wolfbiter) [inline mixes/containers/mix.js]: TODO

File: channels/reducer.js [hunk 13]
```diff
@@ -63,26 +64,6 @@ function createReducer (config) {
...
-            id: channelId,
-            type: CHANNEL_TYPE_PRIMARY_TRACK,
-            clipIds: [clipId] // TODO: maybe setClipChannel in future?
-          }, attrs))),
-          Effects.constant(setChannelParent({ parentChannelId, channelId }))
-        ])
-      }
-
-      return loop(state, Effects.constant(createSample({ file, effectCreator })))
-    },
```
Comment (wolfbiter) [inline channels/reducer.js:85]: moved this function to be next to `swapPrimaryTracks`

File: channels/reducer.js [hunk 14]
```diff
@@ -117,6 +98,49 @@ function createReducer (config) {
...
+        ...state,
+        dirty: [...state.dirty, sourceId, targetId],
+        records: {
+          ...state.records,
+          [sourceId]: merge(state.records[sourceId], {
+            startBeat: target.startBeat
+          }),
+          [targetId]: merge(state.records[targetId], {
+            startBeat: source.startBeat
+          })
```
Comment (wolfbiter) [inline channels/reducer.js:141]: i couldnt loop `updateChannel` here because i needed the store to be updated for the next `swapPrimaryTracks` effect. hopefully wont be runnign into this often, and if i am, should figure out a good pattern

File: styles/file-drop.css [hunk 15]
```diff
@@ -16,3 +16,22 @@
     background-color: rgba(0,155,0,0.5);
   }
 }
+
+.primary-track-table-drag-over {
```
Comment (wolfbiter) [inline styles/file-drop.css:20]: @ahdinosaur starting to need some styling. we should setup that infrastructure next time around

File: package.json [hunk 16]
```diff
@@ -54,7 +54,11 @@
   "devDependencies": {
     "ava": "^0.17.0",
     "chokidar": "^1.6.1",
+    "classnames": "^2.2.5",
     "electron-connect": "^0.6.1",
+    "react-dnd": "^2.1.4",
+    "react-dnd-html5-backend": "^2.1.2",
+    "react-table": "^4.4.1",
```
Comment (wolfbiter) [inline package.json:61]: move all these to normal dependencies

File: layout/container.js (hunk 1 above)
Comment (ahdinosaur) [reply layout/container.js:23]: nice, i like it.

File: styles/file-drop.css (hunk 15 above)
Comment (ahdinosaur) [reply styles/file-drop.css:20]: okay, keen.

**Review by ahdinosaur — APPROVED** [review]
